npep/npep-122.md

```diff
@@ -0,0 +1,194 @@
+    may own more than 1 Namespace.
+
+#### Story 4.3: Allow internal connections for tenants
+    
+    As a cluster admin, I want to build tenants in my cluster and always allow connections inside one tenant.
+    At the same time I want to setup an overridable deny-all policy to protect namespaces by default.
+    This policy should make sure internal connectivity for a tenant is always allowed, in case there are
+    lower-priority deny rules.
+
+#### Story 4.4: Tenants interaction with (B)ANP
+
+    As a cluster admin, I want to be able to setup policies with higher and lower priority than tenancy policy.
+    Higher priority to allow/deny some inter-tenant connections, for example, allow ingress traffic from the
+    monitoring namespace without including that namespace to every tenant. Lower priority to allow/deny
+    connections that are not a part of tenant policy, for example, allow internal tenant connections with tenancy
+    and deny-all policy with lower priority.
+
+#### Story 4.5: Ensure every namespace in the cluster is a part of some tenant
+
+    As a cluster admin, I want to implement security zones tenants, and deny connections between different zones.
+    I want to use a label e.g. `security-zone` to build tenants, and I want to make sure every namespace in the cluster
+    is a part of some tenant/security zone. This way, I can be sure that even if the `security-zone` 
+    label is not assigned to the tenant for some reason, tenancy rules will still apply and security zones won't be
+    violated.
+
+#### What I couldn't figure our user stories for
+
+- Skip action
+- Ports *[]AdminNetworkPolicyPort
+- Multiple tenant policies (e.g. based on different labels?)
+
+### Existing API
+
+AdminNetworkPolicy has the “SameLabels” and “NotSameLabels” fields to support the use cases involving tenancy. For example:
+
+**Use case**
+
+Traffic should be disallowed by default between namespaces owned by different users (defined by labels i.e `user=foo` or `user=bar`).
+
+**Policy**
+```
+kind: BaselineAdminNetworkPolicy
+apiVersion: policy.networking.k8s.io/v1alpha1
+metadata:
+  name: user-based-tenancy
+spec:
+  subject:
+    namespaces:
+      matchExpressions:
+        - key: user
+          operator: Exists
+  ingress:
+    - action: Deny
+      from:
+        - namespaces:
+            notSameLabels:
+              - user
+```
+
+**Meaning**
+
+In namespaces that have a “user” label, by default, deny ingress from namespaces that have a different “user” label.
+
+There are two major problems with this implementation of tenancy as it pertains to the user stories.
+
+First, There is no explicit definition of "tenancy" anywhere. The administrator has an idea of 
+"tenants are defined by the user label", but that's only true because this particular ANP happens to include that
+particular rule, and there's no way to find the ANP(s) that defines tenancy if you don't already know what they are.
```

Comment:
While this is true, I'm not sure this is a huge problem unique to tenancy. I could make a parallel argument for something like egress rules: There's no way to find the ANP(s) that define egress if you don't already know what they are.

I would argue that this is a problem easily solved by a good naming scheme (e.g. call my ANP "tenant-policy" or "namespace-isolation"). Easy discoverability of a particular type of rule shouldn't be a top-priority for our design, in my opinion.

Reply:
Very valid point, I don't think this is the major problem, but more of an introduction to the following points, just stressing that tenants are different from the other peer types. As you said, if you want to define a tenant, you will probably create a separate ANP with a single tenancy rule and a tenancy-specific name, but that is not dictated by the API, and tenancy may easily be lost in a list of other peers.